Add grammar for task lists

burodepeper opened the pull request.

kevinsawicki commented:
Thanks for adding this, would you mind adding some specs for this? You can find them in the `spec/` directory at the root of the repository.

burodepeper commented:
Sure, but I'm completely unfamiliar with that. Is there an Atom specific quickstart guide on that?

kevinsawicki commented:
Yup, here is the raw guide https://github.com/atom/docs/blob/5d3d00a7c895b66d99a227067d3ecf5b0a20951c/book/03-hacking-atom/sections/03-a-basic-package.asc#running-tests

The hosted guide on https://atom.io/docs is currently down, apologies.

burodepeper commented:
Thanks! I'll look into it

burodepeper commented:
I've added the specs. I've noticed that on line 573, I've added an inline bold thing to try and break it, but it still passes the test. Is that intentional?

kevinsawicki commented on grammars/gfm.cson:
```diff
@@ -867,6 +867,30 @@
+      '2':
+        'name': 'space.gfm'
+      '3':
+        'name': 'text.gfm'
```
You need the following here if you want to parse markdown on the line itself:

```coffee
'patterns': [
      {
        'include': '$self'
      }
    ]
```

This causes it to use all the existing patterns between the start and end patterns to further tokenize it.

kevinsawicki commented on grammars/gfm.cson:
```diff
@@ -867,6 +867,30 @@
+        'name': 'text.gfm'
+  }
+  {
+    'begin': '^\\s*(- \\[x\\])(\\s*)(.*)'
```
I'm not sure you want the `(.*)` at the end of this pattern, if you instead use the `$self` technique that I commented below, it will support parsing bold, italic, etc. in between the `- [x]` and the end of the line.

burodepeper commented:
Thanks for the tips. I'll take a look at them tomorrow.

burodepeper commented:
Thanks again for the tips. I'm more of a _trial-and-error_ kind of guy, than a _first-read-the-manual_ one. I think I understand the `'patterns'` addition. I had it on the wrong level/scope at first.

The `(.*)` is more a personal preference. This way I can visually target the text part of a task without touching the rest. What do you think about extending this behavior to normal (un)ordered lists?

burodepeper commented:
Darn it, already found a bug;
`- [ ] ### Header` actually parses that as a header while it shouldn't.
In `- ### Header` it doesn't.

I am a bit at a loss here. The current lists only mark the `- ` bit as the list, and leaves the rest as blank text inside the containing `source.gfm` I think.

I assumed the following is the behavior you'd want:
```markdown
- Apples
- Bananas
```
becomes something similar to
```html
<ul>
<li>- Apples</li>
<li>- Bananas</li>
</ul>
```
while it actually creates (something similar to)
```html
<ul>
<li>- </li>Apples
<li>- </li>Bananas
</ul>
```

I tried to fix that in my attempt at task lists, but the `patterns` bit treats the content as a new line, and thus headings become valid again.

Since all four variants are lists, what should be the default behavior?

burodepeper commented:
Okay, I've done some digging, and I think I know how to fix my issue with the headers. I need add something like:
```cson
'repository':
  'inline-stuff':
    'patterns': [
      // ...
    ]
```
and then modify the
```cson
'patterns': [
  {
    'include': '#inline-stuff'
  }
]
```

Still interested in what you think about the default behavior for lists

kevinsawicki commented:
> Since all four variants are lists, what should be the default behavior?

After giving this some thought, I think unordered lists and task lists should behave the same.

And since unordered lists currently just parse the `* ` part, would you be okay having task lists do the same and just parse the `* [ ]`? Or were you looking to be able to style the full line differently based on whether it was a task list or not?

burodepeper commented:
> Or were you looking to be able to style the full line differently based on whether it was a task list or not?

Yup. For the placebo-effect of _crossing something off of your list_. Gives it more of a sense of purpose/progress than just deleting the completed items from a list.

I'll parse it in a simpler way for now (basically my first solution), and leave my specific styling preferences for perhaps another time.

burodepeper commented:
Current version produces:

![schermafbeelding 2015-08-18 om 10 39 13](https://cloud.githubusercontent.com/assets/3703530/9325859/e3ce8b5a-4595-11e5-865f-6674fca7a328.png)

burodepeper commented:
I've moved all inline markup to a repository, cleaned things up a bit, and updated the patterns of the task list parsing. Task lists are now tokenized in the same way normal lists are, with the exception for completed tasks (for styling purposes). And all `include '$self'` are replaced by a `include '#inline-markup'`.

burodepeper commented:
![schermafbeelding 2015-08-30 om 17 33 27](https://cloud.githubusercontent.com/assets/3703530/9567975/5836a7c4-4f3d-11e5-8c2f-5caa333be7f3.png)

Found this bug, is in the master repo, will fix it tomorrow.

burodepeper commented:
Closing this because I need a clean `master` branch and this has gone stale anyway.
